### ISSUE TITLE
Linear trend in MPI_Reduce time for the same message size

### ISSUE THREAD

**Issue by shiningsurya** [open-mpi/ompi#6983]
Thank you for looking at this issue!!

### What version of Open MPI are you using? (e.g., v1.10.3, v2.1.0, git branch name and hash, etc.)
v4.0.1

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Installed from source

### Please describe the system on which you are running

12 node cluster interconnected with infiniband.

* Operating system/version: Linux
* Computer hardware: Each node has Intel Xeon with 2 numa nodes.
* Network type: There's a 1Gbps ethernet and 40Gbps Infiniband.

-----------------------------

## Details of the problem

_not a problem but having hard time explaining the timing of MPI_Reduce call_

The 12 node cluster is used in a pipeline which is run real time. The main crux of the the MPI application is an `MPI_Reduce` step with `count=2621440` and `datatype=MPI_FLOAT`. So the total memory requirement is `memory=10485760 bytes (~10MB)`. The message size remains the same. 

There are 16 processes launched with few nodes running 2 processes. 

*MPI_Reduce time increases linearly in time for the same message size*. 

See the graphs in the [drive folder](https://drive.google.com/open?id=1Yv8Q3fW99MS-1HiB5uQYAw9aogK3D2Gv). 
`Reduce_coadd` is the MPI_Reduce call time. `DataLoop` is the time taken for one iteration of the pipeline. `readbuf` is the state of the ring buffer. 

*the linear trend is evident in the `Reduce_coadd` plot.

mca file
```shell
--mca btl openib, vader, self, tcp
-- mca btl_openib_allow_ib 1
--mca coll_tuned_reduce_algorithm 7
--mca coll_tuned_reduce_algorithm_segmentsize 200
--mca coll_tuned_priority 100
--mca coll_tuned_barrier_algorithm 3
--mca coll_sync_priority 100
--mca coll_sync_barrier_before 3
```
I would expect the MPI_Reduce to have the same time in each iteration. There is nothing different in any of the iteration.  

Please help me understand this more. 
Thank you for your time

**Comment by ggouaillardet** [comment]
If you want to force a given algorithm, you also need to pass `--mca coll_tuned_use_dynamic_rules true` to your `mpirun` command line.

You should first figure out whether the time is spent in the communication/computation part of `MPI_Reduce()` or inside the implicit barrier.
Do not use `coll/sync`, wrap MPI_Reduce(), add a `MPI_Barrier()` before and after `PMPI_Reduce()` and time everything.

**Comment by shiningsurya** [comment]
Hi 

Thanks for the reply. 
It pushed me in the right direction. 

I followed your suggestion to `coll_tuned_use_dynamic_rules=1` and `coll_tuned_priority=100`. 

_The linear trend is still there but it's so slow that it isn't affecting us anymore. We have a lot of headroom for the linear trend to start crashing our realtime pipeline._

But occasionally, at times, one of a `barrier/small-msg-reduce` takes really long time. 
We were able to trace the origin of that delay to `tcp` links. So, we removed it from the `btl` module. 

Things seem stable now, so I am closing the issue. 

Thanks again!
